### PR TITLE
A (not-quite drop-in) replacement for LMDBStore.

### PR DESCRIPTION
**Pull request by CjS77**
It's presented in parallel so that users can migrate at will.

This version is much more ergonomic and presents a less leaky abstraction than the current implementation.

Key changes:

* There's no more "current db". You can get a handle to a DB and work with it for as long as you like.
* no more get_raw. Everything is abstracted away to get<T> and insert<T> where T: (De)Serialize.
* iterators! (almost). the for_each function offers the same functionality and works just like Vec::for_each.
* len() gives the size of the DB (table)
* log_info() can be used to periodically dump useful stats to the log.
* The docstring outlines the journey to reaching this design and the tradeoffs made wrt zero copy and being opinionated about serialisation.

## Types of changes

* [ ] Bug fix (non-breaking change which fixes an issue)
* [ ] New feature (non-breaking change which adds functionality)
* [ ] Breaking change (fix or feature that would cause existing functionality to change)
* [x] Feature refactor (No new feature or functional changes, but performance or technical debt improvements)
* [ ] New Tests
* [ ] Documentation

## Checklist:

* [x] I'm merging against the `development` branch
* [x] I ran `cargo-fmt --all` before pushing
* [x] My change requires a change to the documentation.
* [x] I have updated the documentation accordingly.
* [x] I have added tests to cover my changes.

### CODE REVIEW

File: infrastructure/storage/src/lmdb_store/store.rs [hunk 1]
```diff
@@ -0,0 +1,586 @@
...
+    /// Sets the size of the database, in MB.
+    /// The actual memory will only be allocated when #build() is called
+    pub fn set_mapsize(mut self, size: usize) -> LMDBBuilder {
+        self.db_size_mb = size;
+        self
+    }
+
+    /// Sets the maximum number of dbs (tables) in the environment. If this value is less than the number of DBs that
+    /// are added to the environmet, this value will be ignored.
+    pub fn set_max_dbs(mut self, size: usize) -> LMDBBuilder {
```
Comment (SWvheerden) [inline infrastructure/storage/src/lmdb_store/store.rs]: perhaps make the function: `set_max_db_size`
This is not much larger but much clearer as to what the function does

File: infrastructure/storage/src/lmdb_store/store.rs [hunk 2]
```diff
@@ -0,0 +1,586 @@
...
+        self.path = path.into();
+        if !self.path.ends_with("/") {
+            self.path += "/";
+        }
+        self
+    }
+
+    /// Sets the size of the database, in MB.
+    /// The actual memory will only be allocated when #build() is called
+    pub fn set_mapsize(mut self, size: usize) -> LMDBBuilder {
```
Comment (SWvheerden) [inline infrastructure/storage/src/lmdb_store/store.rs]: Would this not be more clear to just make this: `set_db_size`

File: infrastructure/storage/tests/lmdb.rs [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+            first,
+            last,
+            email,
+            male,
+            ip,
+        })
+    }
+}
+
+impl ExtendBytes for User {
```
Comment (SWvheerden) [inline infrastructure/storage/tests/lmdb.rs]: you should be able to derive this using the extendbytes macro

File: infrastructure/storage/tests/lmdb.rs [hunk 4]
```diff
@@ -0,0 +1,247 @@
+// Copyright 2019. The Tari Project
```
Comment (SWvheerden) [inline infrastructure/storage/tests/lmdb.rs]: All the tests can be improved by trying to read the values again. The tests currently assume that the lmdb will fail in such a way that an error is thrown if it fails.

File: infrastructure/storage/src/lmdb_store/store.rs (hunk 1 above)
Comment (CjS77) [reply infrastructure/storage/src/lmdb_store/store.rs]: Size suggests a size in MB. This function sets the number of databases that are allowed.

If it's unclear, a better name could be `set_db_capacity`

File: infrastructure/storage/src/lmdb_store/store.rs (hunk 2 above)
Comment (CjS77) [reply infrastructure/storage/src/lmdb_store/store.rs]: Sure. Can change that.

File: infrastructure/storage/tests/lmdb.rs (hunk 4 above)
Comment (CjS77) [reply infrastructure/storage/tests/lmdb.rs]: I'm not clear on what you're saying here. Could you elaborate?

File: infrastructure/storage/tests/lmdb.rs (hunk 3 above)
Comment (CjS77) [reply infrastructure/storage/tests/lmdb.rs]: Does the macro handle the last line?
`buf.extend_from_slice(&self.ip.to_string().as_bytes());`

File: infrastructure/storage/src/lmdb_store/store.rs (hunk 1 above)
Comment (SWvheerden) [reply infrastructure/storage/src/lmdb_store/store.rs]: set_total_databases would perhaps be the best option

File: infrastructure/storage/tests/lmdb.rs (hunk 3 above)
Comment (SWvheerden) [reply infrastructure/storage/tests/lmdb.rs]: Currently I dont think it does. 
But I would add a cfg flag like it currently does with chrono to support Ipv4Addr

File: infrastructure/storage/tests/lmdb.rs (hunk 4 above)
Comment (SWvheerden) [reply infrastructure/storage/tests/lmdb.rs]: The following tests only write data to the db:
-  fn multi_thread() 
- fn multi_thread_writes()

After the data was written there is no check to ensure that the data is actually in the db. All that is actually tested is that write does not panic, or return an error.

File: infrastructure/storage/tests/lmdb.rs (hunk 3 above)
Comment (CjS77) [reply infrastructure/storage/tests/lmdb.rs]: Ok, so I'll leave it handrolled for now

File: infrastructure/storage/tests/lmdb.rs (hunk 4 above)
Comment (CjS77) [reply infrastructure/storage/tests/lmdb.rs]: It is checked in multi_thread, actually, but good catch on mtw -- added

**Pull request merged**